virtual_rainforest/models/animals/animal_cohorts.py

```diff
@@ -126,16 +126,16 @@ def metabolize(self, dt: timedelta64) -> None:
         energy_needed = self.metabolic_rate * float((dt / timedelta64(1, "s")))
         self.stored_energy -= min(self.stored_energy, energy_needed)
 
-    def excrete(self, soil: Pool, consumed_energy: float) -> None:
-        """The function to transfer waste energy from an animal cohort to the soil.
+    def excrete(self, excrement_pool: Pool, consumed_energy: float) -> None:
```

Comment:
I'm wondering whether, instead of taking a `Pool` argument to modify, this function could just return the amount of waste energy and then the caller could update a `Pool` if it wants. What do you reckon?

Reply:
I'm not sure because I don't think the step to calculate waste energy would ever be invoked for any reason but to update the excrement pool, so I don't feel like there's a value in having it be a separate function?

Reply:
I was thinking that it would make it more modular and testable. The problem with passing `excrement_pool` in and modifying it is that `excrete()` then depends on the state of `excrement_pool` in addition to `consumed_energy` and various attributes of `self`.

On a related note, I've had a look and it seems there's not a test for `excrete()` as well as various other smaller functions. (I know this isn't your doing in particular -- I'm not singling you out!) There are tests for higher-level functions, so the test coverage looks good, but this isn't the best practice for unit testing. Ideally you want separate tests for smaller functions, checking that certain inputs give the expected outputs, then you can (often) just test that higher-level functions call the lower-level ones in the correct way with mocking.

I've noticed that you've just changed `excrete()` to split the energy between two pools. I'm not entirely sure about the way you've implemented this. Is `decay_fraction` something you expect to change between calls to `excrete()`? If so, it should be settable via `forage_communities()` too. If not, then it should be set once for the `AnimalCohort` in `__init__`.

Reply:
PS -- I have reviewed large chunks of this code without twigging that these smaller functions aren't being tested, so it's partly on me. I would suggest that you guys adopt a different approach in future though.

Reply:
I took a look and the only missing tests in the animal module appear to for the newly implemented `prey_group_selection` and `migrate_community` methods, which I'm still fiddling with. Some of these tests may need to be expanded in functionality but I can't find a significant hole in coverage. Perhaps Jacob removed some tests from this branch while he was changing the forms.

Reply:
Yeah I'm pretty sure that `excrete()` is tested (within the `TestAnimalCohort` testing class) as I had to modify the test when I altered the `excrete()` function

Reply:
On the `decay_fraction` question, for the meantime this is not something that should change between calls to `excrete()`. In future, we _might_ want to change the model such that it can change between (due to e.g. temperature changes). I guess it's not sensible to hedge against this potential future use and to just set it in the `__init__`?

Reply:
My bad @TaranRallings. You're quite right. I totally missed it, but it does seem like there are plenty of tests.